File: js/entities/world/Coin.js
// js/entities/world/Coin.js
import { Entity } from '../Entity.js';
import { state } from '../../state.js';
import { COIN_RADIUS, COIN_ORBIT_OFFSET } from '../../constants.js';
import { Vector2 } from '../../vector2.js';
import { Container, Sprite, Texture } from 'pixi.js';

// Shared across EVERY coin — COIN_RADIUS is a fixed global constant
// (not varied per-planet or per-coin), so every coin's body/highlight
// are pixel-for-pixel identical; only position and rotation differ per
// instance. Baked once here rather than as live per-instance Graphics
// objects, which is what this file originally used and turned out to
// be the actual cause of a real, measured framerate drop once coins
// were added — confirmed directly against Pixi's own performance
// documentation ("Using 100s of graphics complex objects can be slow,
// in this instance use sprites (you can create a texture)"), not just
// assumed: with potentially thousands of coins alive at once across a
// full 3x3 cell neighborhood, two live Graphics objects EACH is
// precisely the "hundreds of Graphics" case that guidance warns about.
// A shared, pre-baked texture + cheap Sprite per instance is the
// officially-recommended fix, and what every other entity in this
// migration already does — this was the one file that deviated from
// that pattern, not a new technique introduced here.
let coinBodyTexture = null;
function getCoinBodyTexture() {
  if (coinBodyTexture) return coinBodyTexture;
  const rx = COIN_RADIUS * 0.65;
  const ry = COIN_RADIUS;
  const strokeWidth = 2;
  const pad = strokeWidth; // margin so the rim stroke doesn't get clipped at the canvas edge
  const canvas = document.createElement('canvas');
  canvas.width = (rx + pad) * 2;
  canvas.height = (ry + pad) * 2;
  const ctx = canvas.getContext('2d');
  ctx.translate(rx + pad, ry + pad);
  ctx.beginPath();
  ctx.ellipse(0, 0, rx, ry, 0, 0, Math.PI * 2);
  ctx.fillStyle = '#ffd700'; // gold
  ctx.fill();
  ctx.strokeStyle = 'rgba(184,134,11,0.8)'; // darker gold rim
  ctx.lineWidth = strokeWidth;
  ctx.stroke();
  coinBodyTexture = Texture.from(canvas);
  return coinBodyTexture;
}

let coinHighlightTexture = null;
function getCoinHighlightTexture() {
  if (coinHighlightTexture) return coinHighlightTexture;
  const r = COIN_RADIUS * 0.4;
  const canvas = document.createElement('canvas');
  canvas.width = r * 2;
  canvas.height = r * 2;
  const ctx = canvas.getContext('2d');
  ctx.beginPath();
  ctx.arc(r, r, r, 0, Math.PI * 2);
  ctx.fillStyle = 'rgba(255,255,255,0.8)';
  ctx.fill();
  coinHighlightTexture = Texture.from(canvas);
  return coinHighlightTexture;
}

export class Coin extends Entity {
  constructor(planet) {
    super();
    this.planet = planet;
    this.angularSpeed = (Math.random() - 0.5) * 0.04;
    this.angle = 0;
    this.radius = COIN_RADIUS;
    this.orbitRadius = planet.radius + COIN_ORBIT_OFFSET;

    // Rounded-rect planets don't have a single "angle" — they use the
    // same arc-length perimeter-walk parameterization the player uses
    // to walk their surface, just always drifting (never player-
    // controlled) and pushed further out (COIN_ORBIT_OFFSET instead of
    // PLAYER_RADIUS).
    this.orbitOffset = COIN_ORBIT_OFFSET;
    this.arcPos = planet.isRoundedRect ? Math.random() * planet.getPerimeter() : 0;
    this.arcSpeed = (Math.random() - 0.5) * 2; // px/frame, comparable visual speed to the circular case

    // Elliptical, vertically-long dimensions for the new Pixi rendering
    // below (see createPixiSprite) — classic Mario-coin proportions
    // (narrower than tall), rather than draw()'s own plain circle.
    // Deliberately separate from this.radius, which stays exactly as
    // it was and keeps driving draw()'s own untouched Canvas2D
    // fallback, plus the actual pickup-collision radius elsewhere
    // (CollisionSystem.js's own handleCoinCollisions) — this is a
    // purely visual change, not a hitbox change.
    this.radiusX = this.radius * 0.65;
    this.radiusY = this.radius;

    this.updatePosition();
  }

  updatePosition() {
    if (this.planet.isRoundedRect) {
      const world = this.planet.worldPointAtArcPosition(this.arcPos, this.orbitOffset);
      this.pos.x = world.point.x;
      this.pos.y = world.point.y;
    } else {
      this.pos.x = this.planet.pos.x + Math.cos(this.angle) * this.orbitRadius;
      this.pos.y = this.planet.pos.y + Math.sin(this.angle) * this.orbitRadius;
    }
  }

  update() {
    if (this.planet.isRoundedRect) {
      this.arcPos += this.arcSpeed * state.timeScale;
    } else {
      this.angle += this.angularSpeed * state.timeScale;
    }
    this.updatePosition();
  }

  draw() {
    const ctx = state.ctx;
    ctx.beginPath();
    ctx.arc(this.pos.x, this.pos.y, this.radius, 0, Math.PI * 2);
    ctx.fillStyle = 'gold';
    ctx.fill();
    ctx.save();
    ctx.translate(this.pos.x, this.pos.y);
    ctx.rotate(Date.now() * 0.01);
    ctx.beginPath();
    ctx.arc(-this.radius*0.4, -this.radius*0.4, this.radius*0.4, 0, Math.PI*2);
    ctx.fillStyle = 'rgba(255,255,255,0.8)';
    ctx.fill();
    ctx.restore();
  }

  // Lazily builds the coin's Pixi representation — a body Sprite and a
  // highlight Sprite, both wrapping the shared textures above, rather
  // than either being its own live Graphics object (see this file's
  // own header comment on why that was the actual performance bug).
  // draw()'s own rotating highlight worked by rotating the CANVAS'S
  // LOCAL COORDINATE SYSTEM around the coin's center, not by moving
  // the highlight's own position; the Pixi equivalent of that isn't a
  // redraw at all, it's a Container (highlightPivot below) whose own
  // .rotation gets set every frame in updatePixiSprite — the same
  // "rotate the local space, not the shape" idea, just expressed as a
  // transform instead of a redraw, so there's nothing to rebuild every
  // frame for either sprite, only two numbers (position, rotation) to
  // update on the whole coin each frame.
  createPixiSprite(coinLayer) {
    if (this.pixiContainer) return;

    const container = new Container();

    const body = new Sprite(getCoinBodyTexture());
    body.anchor.set(0.5, 0.5);

    const highlightPivot = new Container();
    const highlight = new Sprite(getCoinHighlightTexture());
    highlight.anchor.set(0.5, 0.5);
    // The highlight texture itself is a plain, centered circle with no
    // offset baked in — this reproduces the same off-center placement
    // (-radiusY*0.4, -radiusY*0.4) the original Graphics version baked
    // directly into its own circle() call, applied here as the
    // sprite's own position within its rotating parent instead.
    highlight.position.set(-this.radiusY * 0.4, -this.radiusY * 0.4);
    highlightPivot.addChild(highlight);

    container.addChild(body, highlightPivot);
    coinLayer.addChild(container);

    this.pixiContainer = container;
    this.pixiHighlightPivot = highlightPivot;
  }

  // Called every frame in place of draw() once migrated.
  updatePixiSprite(coinLayer) {
    if (!this.pixiContainer) this.createPixiSprite(coinLayer);

    this.pixiContainer.position.set(this.pos.x, this.pos.y);
    this.pixiHighlightPivot.rotation = Date.now() * 0.01;
  }

  // Called from PixiStage.js's own sync function when this coin is
  // removed (picked up, or its planet was culled/despawned — see
  // CollisionSystem.js's own handleCoinCollisions and worldGen.js's
  // own cullDistantObjects). No texture:true here — both sprites wrap
  // the SHARED textures above (every coin's body/highlight are
  // pixel-for-pixel identical), so only the sprite/container instances
  // themselves get destroyed; destroying the shared texture itself
  // would break every other still-alive coin's own rendering.
  destroyPixiSprite() {
    if (this.pixiContainer) {
      this.pixiContainer.destroy({ children: true });
      this.pixiContainer = null;
      this.pixiHighlightPivot = null;
    }
  }
}

File: lua/entities/Coin.lua
-- lua/entities/Coin.lua
--
-- Port of js/entities/world/Coin.js
-- Coins orbit a planet and can be collected by the player.

local state = require("lua.state")
local Vector2 = require("lua.vector2")
local constants = require("lua.constants")

local Coin = {}
Coin.__index = Coin

local COIN_RADIUS = constants.COIN_RADIUS or 12
local COIN_ORBIT_OFFSET = constants.COIN_ORBIT_OFFSET or 28

function Coin.new(planet)
  local self = setmetatable({}, Coin)

  self.planet = planet
  self.radius = COIN_RADIUS
  self.orbitRadius = planet.radius + COIN_ORBIT_OFFSET
  self.orbitOffset = COIN_ORBIT_OFFSET

  self.angularSpeed = (math.random() - 0.5) * 0.04
  self.angle = math.random() * math.pi * 2

  -- For rounded-rect planets (arc-length walk)
  self.arcPos = 0
  self.arcSpeed = (math.random() - 0.5) * 2
  if planet.isRoundedRect and planet.getPerimeter then
    self.arcPos = math.random() * planet:getPerimeter()
  end

  self.pos = Vector2.new(0, 0)
  self:updatePosition()

  return self
end

function Coin:updatePosition()
  local planet = self.planet
  if planet.isRoundedRect and planet.worldPointAtArcPosition then
    local world = planet:worldPointAtArcPosition(self.arcPos, self.orbitOffset)
    self.pos.x = world.point.x
    self.pos.y = world.point.y
  else
    self.pos.x = planet.pos.x + math.cos(self.angle) * self.orbitRadius
    self.pos.y = planet.pos.y + math.sin(self.angle) * self.orbitRadius
  end
end

function Coin:update()
  local ts = state.timeScale or 1
  if self.planet.isRoundedRect then
    self.arcPos = self.arcPos + self.arcSpeed * ts
  else
    self.angle = self.angle + self.angularSpeed * ts
  end
  self:updatePosition()
end

function Coin:draw()
  local x, y = self.pos.x, self.pos.y
  local r = self.radius

  -- Gold body
  love.graphics.setColor(1.0, 0.84, 0.0, 1)
  love.graphics.circle("fill", x, y, r)

  -- Darker gold rim
  love.graphics.setColor(0.72, 0.53, 0.04, 0.8)
  love.graphics.setLineWidth(2)
  love.graphics.circle("line", x, y, r)

  -- Spinning highlight
  local spin = love.timer.getTime() * 10
  love.graphics.push()
  love.graphics.translate(x, y)
  love.graphics.rotate(spin)
  love.graphics.setColor(1, 1, 1, 0.8)
  love.graphics.circle("fill", -r * 0.4, -r * 0.4, r * 0.4)
  love.graphics.pop()

  love.graphics.setColor(1, 1, 1, 1)
  love.graphics.setLineWidth(1)
end

return Coin